Minting script witness refactor

Jimbo4350 opened the pull request:
# Changelog

```yaml
- description: |
    Minting script witness refactor
    The type `ScriptWitnessFiles` makes it difficult to accommodate for specific changes to plutus scripts. 
    As a result we introduce `MintScriptWitnessWithPolicyId` as a first step towards deprecating `ScriptWitnessFiles`. 
    This paves the way to more readable code and allows us to introduce specific changes to the different script types i.e simple vs plutus.

  type:
  # - feature        # introduces a new feature
  - breaking       # the API has changed in a breaking way
  # - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  - refactoring    # QoL changes
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

Additional context for the PR goes here. If the PR fixes a particular issue please provide a [link](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword=) to the issue.

# How to trust this PR

Highlight important bits of the PR that will make the review faster. If there are commands the reviewer can run to observe the new behavior, describe them.

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated. See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] Self-reviewed the diff

<!--
### Note on CI ###
If your PR is from a fork, the necessary CI jobs won't trigger automatically for security reasons.
You will need to get someone with write privileges. Please contact IOG node developers to do this
for you.
-->

smelc committed:
Use new inject function instead of the XToY era functions

Jimbo4350 committed:
Add Cardano.CLI.Plutus.Minting module

Define the data definition CliMintScriptRequirements
This type makes it clearer that we require the policy id for transaction
construction when using a minting script

Jimbo4350 committed:
Refine plutusP to return AnyPlutusScriptVersion instead of
AnyScriptLanguage
Replace ScriptWitnessFiles WitCtxMint with CliMintScriptRequirements

Jimbo4350 committed:
Modify PlutusReferenceScriptWitnessFiles to accept
AnyPlutusScriptVersion instead of AnyScriptLanguage

Jimbo4350 committed:
Implement readFileSimpleScript and readFilePlutusScript

Refactor readScriptWitness and eliminate invalid states

Factor out fromSomeTypeSimpleScript and fromSomeTypePlutusScripts

fromSomeTypePlutusScripts should automatically be updated as soon as the
Enum AnyPlutusScriptVersion instance is updated in cardano-api

Jimbo4350 committed:
Replace ScriptWitnessFiles WitCtxMint with CliMintScriptRequirements

carbolymer committed:
Remove use of experimental api in transaction balancing

Jimbo4350 committed:
Propagate MintScriptWitWithPolId era

Jimbo4350 committed:
Remove gatherMintingWitnesses because by default we require a PolicyId

Jimbo4350 committed:
Add TxCmdCliScriptWitnessError to TxCmdError
Add ScriptDecodeUnknownPlutusScriptVersion to ScriptDecodeError

carbolymer commented on cardano-cli/src/Cardano/CLI/Plutus/Minting.hs:
```diff
@@ -0,0 +1,200 @@
+  , MintScriptWitWithPolId (..)
+  , createOnDiskSimpleOfPlutusScriptCliArgs
+  , createOnDiskSimpleReferenceScriptCliArgs
+  , createOnDiskPlutusReferenceScriptCliArgs
```
I think we should switch to either using `OnDiskXXXXThing` or `FileXXXXThing`  names convention for the consistency, eventually. It would make navigating codebase easier.

Jimbo4350 replied:
What about:

```
   createSimpleOrPlutusScriptFromCliArgs
   createSimpleReferenceScriptFromCliArgs
   createPlutusReferenceScriptFromCliArgs
```

carbolymer replied:
That's fine I think. I had in mind the type names as well.

carbolymer commented on cardano-cli/src/Cardano/CLI/Plutus/Minting.hs:
I don't think this module is in the right place. Can we split it into two:

- `cardano-cli/src/Cardano/CLI/Types/Plutus.hs` with `MintScriptWitWithPolId`, `CliMintScriptRequirements` and `createOnDiskXXX` functions
- `cardano-cli/src/Cardano/CLI/Read/Plutus.hs` with `readMintScriptWitness`  and `CliScriptWitnessError`

?

I'm not a fan of the current module structure, but it never felt like a good moment to turn everything upside down. 

For the moment I think we should follow it rather than introducing different changes.

I propose the following change:
https://gist.github.com/carbolymer/63b23d0df940c61d1501a6763d86134c

I'm not insisting on keeping `EraBased` if we have a better alternative here.

Thoughts?

Jimbo4350 replied:
100% agree with the new structure and the module split.

So then based on what you suggested we would have:
```
Cardano/CLI
├── Byron
├── Commands
├── EraBased
│   └── Plutus
│       └── Read.hs
│       └── Types.hs
│     
│           
```

carbolymer replied:
`├── Commands` could be organised by domains too I think.

carbolymer commented on cardano-cli/src/Cardano/CLI/Plutus/Minting.hs:
```diff
@@ -0,0 +1,200 @@
+-- in order to correctly construct the transaction.
+data MintScriptWitWithPolId era
+  = MintScriptWitWithPolId
+  { mswPolId :: PolicyId
```
```suggestion
data MintScriptWitWithPolicyId era
  = MintScriptWitWithPolicyId
  { mswPolicyId :: PolicyId
```
I don't think we're gaining much here by shortening the names here

carbolymer commented on cardano-cli/src/Cardano/CLI/Plutus/Minting.hs:
```diff
@@ -0,0 +1,200 @@
+sbeToSimpleScriptLangInEra ShelleyBasedEraMary = SimpleScriptInMary
+sbeToSimpleScriptLangInEra ShelleyBasedEraAlonzo = SimpleScriptInAlonzo
+sbeToSimpleScriptLangInEra ShelleyBasedEraBabbage = SimpleScriptInBabbage
+sbeToSimpleScriptLangInEra ShelleyBasedEraConway = SimpleScriptInConway
```
- [x] I'll upstream this to `Cardano.Api.Script`.

carbolymer replied:
done in https://github.com/IntersectMBO/cardano-api/pull/663

carbolymer commented on cardano-cli/src/Cardano/CLI/Read.hs:
```diff
@@ -565,20 +568,118 @@ deserialiseScriptInAnyLang bs =
+        "PlutusScriptV1" ->
+          case deserialiseFromTextEnvelopeAnyOf [teTypes (AnyPlutusScriptVersion PlutusScriptV1)] te of
+            Left err -> Left (PlutusScriptDecodeTextEnvelopeError err)
+            Right script -> Right script
```
this can be simplified too

carbolymer commented on cardano-cli/src/Cardano/CLI/Read.hs:
```diff
@@ -565,20 +568,118 @@ deserialiseScriptInAnyLang bs =
+deserialisePlutusScript bs =
+  case deserialiseFromJSON AsTextEnvelope bs of
+    Left err -> Left $ PlutusScriptJsonDecodeError err
+    Right te ->
```
```suggestion
  te <- first PlutusScriptJsonDecodeError $ deserialiseFromJSON AsTextEnvelope bs
```

carbolymer commented on cardano-cli/src/Cardano/CLI/Plutus/Minting.hs:
```diff
@@ -0,0 +1,200 @@
+  -> PolicyId
+  -> CliMintScriptRequirements
+createOnDiskPlutusReferenceScriptCliArgs txin scriptVersion scriptData execUnits polid =
+  OnDiskPlutusRefScript $ PlutusRefScriptCliArgs txin scriptVersion scriptData execUnits polid
```
Can you put those definitions below the type `CliMintScriptRequirements`? It makes hard to read when you have to jump up and down from definition to definiton to follow the call/type hierarchy.

carbolymer commented on cardano-cli/src/Cardano/CLI/Plutus/Minting.hs:
```diff
@@ -0,0 +1,200 @@
+  | OnDiskSimpleRefScript SimpleRefScriptCliArgs
+  | OnDiskPlutusRefScript PlutusRefScriptCliArgs
+  deriving Show
+
```
Same like my previous comment, can you put this definition near the top of the module, before any of its "dependencies"?

(also applies to other types here, I don't want to highlight the whole file for that purpose)

carbolymer reviewed (commented):
LGTM after first pass 👍🏻 Nice work.

Jimbo4350 committed:
Review rename and refactor suggestions

Jimbo4350 committed:
Module split and rename suggestions

carbolymer commented on cardano-cli/src/Cardano/CLI/Read.hs:
```diff
@@ -565,20 +568,131 @@ deserialiseScriptInAnyLang bs =
+    if v == show lang
+      then case deserialiseFromTextEnvelopeAnyOf [teTypes (AnyPlutusScriptVersion lang)] tEnv of
+        Left err -> Left (PlutusScriptDecodeTextEnvelopeError err)
+        Right script -> Right script
```
```suggestion
      then 
        first PlutusScriptDecodeTextEnvelopeError $
          deserialiseFromTextEnvelopeAnyOf [teTypes (AnyPlutusScriptVersion lang)] tEnv
```

carbolymer committed:
Bump cardano-api-10.3

Jimbo4350 committed:
Add separate error modules:
- Cardano.CLI.Types.Errors.PlutusScriptDecodeError
- Cardano.CLI.Types.Errors.ScriptDataError

Jimbo4350 committed:
With the implementation and use MintScriptWitnessWithPolicyId era to
handle all minting scripts (simple and plutus) we no longer have to
accomodate for the PolicyId in the constructors
PlutusReferenceScriptWitnessFiles and SimpleReferenceScriptWitnessFiles

This is evidenced by the diff of this commit

The goal is to deprecate ScriptWitnessFiles and replace it with a
collection of types for the different script purposes. The first example
of this is MintScriptWitnessWithPolicyId era

Jimbo4350 committed:
Update createTxMintValue and propagate the removal of Maybe PolicyId
from PReferenceScript and SReferenceScript constructors

Co-authored-by: Mateusz Gałażyn <mateusz.galazyn@gmail.com>

Jimbo4350 committed:
Formatting fixes

carbolymer commented on cardano-cli/src/Cardano/CLI/Types/Common.hs:
```diff
@@ -398,11 +399,7 @@ type ScriptRedeemerOrFile = ScriptDataOrFile
--- rid of Maybe (ScriptWitnessFiles). This will be clearer
--- in conveying that we either expect a script witness
--- or a key witness is provided at the signing stage.
+-- NB: This is in the process of being deprecated
```
I'm not a fan of vague comments. It will be meaningless for anyone not familiar with the work.
 Can you add a sentence how the replacement should/will look like?

smelc commented on cardano-cli/src/Cardano/CLI/Types/Errors/PlutusScriptDecodeError.hs:
```diff
@@ -0,0 +1,34 @@
+    PlutusScriptDecodeTextEnvelopeError err ->
+      prettyError err
+    PlutusScriptDecodeErrorVersionMismatch version (AnyPlutusScriptVersion v) ->
+      "Version mismatch in code: script version that was read"
+        <> pretty version
+        <> " but tried to decode script version: "
+        <> pshow v
```
```suggestion
      "Version mismatch in code: script version that was read"
        <+> pretty version
        <+> "but tried to decode script version:"
        <+> pshow v
```

smelc commented on cardano-cli/src/Cardano/CLI/Types/Errors/ScriptDataError.hs:
```diff
@@ -0,0 +1,35 @@
+  ScriptDataErrorFile err ->
+    prettyError err
+  ScriptDataErrorJsonParse fp jsonErr ->
+    "Invalid JSON format in file: " <> pshow fp <> "\nJSON parse error: " <> pretty jsonErr
+  ScriptDataErrorConversion fp sDataJsonErr ->
+    "Error reading metadata at: " <> pshow fp <> "\n" <> prettyError sDataJsonErr
+  ScriptDataErrorValidation fp sDataRangeErr ->
+    "Error validating script data at: " <> pshow fp <> ":\n" <> prettyError sDataRangeErr
+  ScriptDataErrorMetadataDecode fp decoderErr ->
+    "Error decoding CBOR metadata at: " <> pshow fp <> " Error: " <> pshow decoderErr
```
```suggestion
    "Invalid JSON format in file:" <+> pshow fp <> "\nJSON parse error:" <+> pretty jsonErr
  ScriptDataErrorConversion fp sDataJsonErr ->
    "Error reading metadata at:" <+> pshow fp <> "\n" <> prettyError sDataJsonErr
  ScriptDataErrorValidation fp sDataRangeErr ->
    "Error validating script data at:" <+> pshow fp <> ":\n" <> prettyError sDataRangeErr
  ScriptDataErrorMetadataDecode fp decoderErr ->
    "Error decoding CBOR metadata at:" <+> pshow fp <+> "Error:" <+> pshow decoderErr
```

smelc commented on cardano-cli/src/Cardano/CLI/Types/Errors/ScriptDecodeError.hs:
```diff
@@ -22,3 +25,5 @@ instance Error ScriptDecodeError where
     ScriptDecodeSimpleScriptError err ->
       "Syntax error in script: " <> prettyError err
+    ScriptDecodeUnknownPlutusScriptVersion version ->
+      "Unknown Plutus script version: " <> pshow version
```
```suggestion
      "Unknown Plutus script version:" <+> pshow version
```

carbolymer commented on cardano-cli/src/Cardano/CLI/Types/Common.hs:
```diff
@@ -398,11 +399,19 @@ type ScriptRedeemerOrFile = ScriptDataOrFile
+-- which says nothing about the script type (simple vs plutus) or script version.
+-- As a result need to separate the different script purposes into
+-- their own separate data definitions where we can make changes specific to that script purpose
+-- more easily without affecting the rest of the api.
```
Beautiful 🤩

carbolymer commented on cardano-cli/src/Cardano/CLI/EraBased/Run/Transaction.hs:
```diff
@@ -1230,9 +1248,9 @@ getAllReferenceInputs
     getReferenceInput sWit =
       case sWit of
-        PlutusScriptWitness _ _ (PReferenceScript refIn _) _ _ _ -> Just refIn
+        PlutusScriptWitness _ _ (PReferenceScript refIn) _ _ _ -> Just refIn
```
This function can be removed, as this functionality is provided by `getScriptWitnessReferenceInput` from cardano-api.
